sdk/identity/identity/src/msal/browserFlows/msalAuthCode.ts

```diff
@@ -3,249 +3,352 @@
+ * Generates a MSAL configuration that generally works for browsers
  * @internal
  */
-export class MSALAuthCode extends MsalBrowser {
-  private loginHint?: string;
-
-  /**
-   * Sets up an MSAL object based on the given parameters.
-   * MSAL with Auth Code allows sending a previously obtained `authenticationRecord` through the optional parameters,
-   * which is set to be the active account.
-   * @param options - Parameters necessary and otherwise used to create the MSAL object.
-   */
-  constructor(options: MsalBrowserFlowOptions) {
-    super(options);
-    this.loginHint = options.loginHint;
-
-    this.msalConfig.cache = {
+function generateMsalBrowserConfiguration(
+  options: MsalBrowserFlowOptions,
+): msalBrowser.Configuration {
+  const tenantId = options.tenantId || DefaultTenantId;
+  const authority = getAuthority(tenantId, options.authorityHost);
+  return {
+    auth: {
+      clientId: options.clientId!,
+      authority,
+      knownAuthorities: getKnownAuthorities(tenantId, authority, options.disableInstanceDiscovery),
+      // If the users picked redirect as their login style,
+      // but they didn't provide a redirectUri,
+      // we can try to use the current page we're in as a default value.
+      redirectUri: options.redirectUri || self.location.origin,
+    },
+    cache: {
       cacheLocation: "sessionStorage",
       storeAuthStateInCookie: true, // Set to true to improve the experience on IE11 and Edge.
-    };
-    this.msalConfig.system = {
+    },
+    system: {
       loggerOptions: {
-        loggerCallback: defaultLoggerCallback(this.logger, "Browser"),
+        loggerCallback: defaultLoggerCallback(options.logger, "Browser"),
         logLevel: getMSALLogLevel(getLogLevel()),
         piiLoggingEnabled: options.loggingOptions?.enableUnsafeSupportLogging,
       },
+    },
+  };
+}
+
+// We keep a copy of the redirect hash.
+const redirectHash = self.location.hash;
+
+/**
+ * Uses MSAL Browser 2.X for browser authentication,
+ * which uses the [Auth Code Flow](https://learn.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-auth-code-flow).
+ * @internal
+ */
+export function createMsalBrowserClient(options: MsalBrowserFlowOptions): MsalBrowserClient {
+  const loginStyle = options.loginStyle;
+  if (!options.clientId) {
+    throw new CredentialUnavailableError("A client ID is required in browsers");
+  }
+  const clientId = options.clientId;
+  const logger = options.logger;
+  const tenantId = resolveTenantId(logger, options.tenantId, options.clientId);
+  const additionallyAllowedTenantIds: string[] = resolveAdditionallyAllowedTenantIds(
+    options?.tokenCredentialOptions?.additionallyAllowedTenants,
+  );
+  const authorityHost = options.authorityHost;
+  const msalConfig = generateMsalBrowserConfiguration(options);
+  const disableAutomaticAuthentication = options.disableAutomaticAuthentication;
+  let app: msalBrowser.IPublicClientApplication;
```

Comment:
nit: maybe move this to line 96 right above the `getApp` call - the proximity will help readers understand that this is just meant to be lazy-loaded via `getApp` - you could also add a comment explaining that we should not use this directly, instead calling `getApp` instead